main/lsp/test/error_reporter_test.cc

```diff
@@ -162,12 +192,129 @@ TEST_CASE("ErrorReporterIgnoresErrorsFromOldEpochs") {
+
+        er.beginEpoch(epoch, move(diagnosticLatencyTimers));
+        Timer::timedSleep(chrono::milliseconds(5), *logger, "debugging");
+        er.pushDiagnostics(epoch, fref, errors, *gs);
+
+        Timer::timedSleep(chrono::milliseconds(5), *logger, "debugging");
+        er.pushDiagnostics(epoch, fref, errors, *gs);
+        er.endEpoch(epoch);
+
+        auto counters = getAndClearThreadCounters();
+        auto counterStateDatabase = CounterStateDatabase(move(counters));
+
+        INFO("Reports first_ and last_diagnostic_latency");
+        CHECK_EQ(1, counterStateDatabase.getTimings("first_diagnostic_latency").size());
+        CHECK_EQ(1, counterStateDatabase.getTimings("last_diagnostic_latency").size());
+
+        // Assert that first_diagnostic_latency's end time is set when pushDiagnostics is called for the first time
+        // on a file and is not updated after
+        INFO("first_diagnostic_latency's end time is not changed in subsequent checks of the same file");
+        auto firstDiagnosticLatency = move(counterStateDatabase.getTimings("first_diagnostic_latency").front());
+        auto firstDiagnosticDuration = firstDiagnosticLatency->end.usec - firstDiagnosticLatency->start.usec;
+        CHECK_LT(chrono::microseconds(firstDiagnosticDuration), chrono::milliseconds(10));
+
+        // Assert that last_diagnostic_latency's end time is updated every time we report errors for a file
+        INFO("last_diagnostic_latency's end time is changed in subsequent checks of the same file");
+        auto lastDiagnosticLatency = move(counterStateDatabase.getTimings("last_diagnostic_latency").front());
+        auto lastDiagnosticDuration = lastDiagnosticLatency->end.usec - lastDiagnosticLatency->start.usec;
+        CHECK_GT(chrono::microseconds(lastDiagnosticDuration), chrono::milliseconds(10));
+    }
+}
+
+TEST_CASE("FirstAndLastLatencyAboutEqualWhenNoErrors") {
+    auto gs = makeGS();
+    auto cs = makeConfig();
+    ErrorReporter er(cs);
+    {
+        core::UnfreezeFileTable fileTableAccess(*gs);
+        auto epoch = 0;
+        auto file = make_shared<core::File>("foo.rb", "foo", core::File::Type::Normal, epoch);
+        auto fref = gs->enterFile(file);
+        auto outputVector = dynamic_pointer_cast<LSPOutputToVector>(cs->output);
+
+        vector<unique_ptr<core::Error>> errors;
+        errors.emplace_back(
+            make_unique<core::Error>(core::Loc(fref, 0, 0), core::ErrorClass{1, core::StrictLevel::True}, "MyError",
+                                     vector<core::ErrorSection>(), vector<core::AutocorrectSuggestion>(), false));
+
+        vector<unique_ptr<Timer>> diagnosticLatencyTimers;
+        diagnosticLatencyTimers.emplace_back(make_unique<Timer>(logger, "last_diagnostic_latency"));
+
+        er.beginEpoch(epoch, move(diagnosticLatencyTimers));
+        Timer::timedSleep(chrono::milliseconds(5), *logger, "debugging");
+        er.pushDiagnostics(epoch, fref, errors, *gs);
```

Comment:
This isn't testing what you want it to test. This call to `pushDiagnostics` will send an LSPMessage since it's sending errors.

I think we need a throwaway epoch that sends an empty error list for this file, then a second epoch where we once again report the empty list for the same file. `ErrorReporter` should not send any LSPMessages, since the error list is still empty.

In-between epochs we should clear counters and LSPMessages so we can isolate what happens during the second epoch.